src/tensor_ops/reduce.rs

```diff
@@ -0,0 +1,48 @@
+use crate::prelude::*;
+
+/// TODO
+pub trait Reduce1<const I: isize>: Tensor<Dtype = f32> {
+    type Reduced: Tensor<Dtype = Self::Dtype, Tape = Self::Tape> + Broadcast1<Self, I>;
+    type DeviceR: Reduce1Axis<Self::Array, <Self::Reduced as HasArrayType>::Array, I>;
```

Comment:
Having a separate device here is necessary because the normal Tensor::Device doesn't have a bound for Reduce1Axis<_, _, I> (and this isn't possible since it doesn't have axis to `I`).

If this new device wasn't specified then any generic function that uses Reduce1 would have to also have T::Device: Reduce1Axis<_, _, I>, which I felt was pretty verbose.